library/core/src/intrinsics.rs

```diff
@@ -1724,6 +1724,25 @@ extern "rust-intrinsic" {
     #[rustc_nounwind]
     pub fn fmaf64(a: f64, b: f64, c: f64) -> f64;
 
+    /// Returns `a * b + c` for `f32` values.
+    ///
+    /// The operation is fused if the code generator determines that target
+    /// instruction set has support for a fused operation, and that the fused
+    /// operation is more efficient than the equivalent, separate pair of mul
+    /// and add instructions.
+    #[rustc_nounwind]
+    #[cfg(not(bootstrap))]
+    pub fn fmuladdf32(a: f32, b: f32, c: f32) -> f32;
```

Comment:
So sounds like the spec is : non-deterministically either fuse or not? But the result is *guaranteed* to be either the fully precise fused computation rounded once, or the fully precise multiplication, rounded, then fully precise addition, rounded?

A spec should not usually mention "codegen". A spec should say what the behavior is. Codegen's job is to implement that.

When I call this multiple times with the same inputs in the same program, am I guaranteed to either always get the fused behavior or always the unfused behavior, or can it differ between different invocations? What if they used different inputs?

Reply:
> So sounds like the spec is : non-deterministically either fuse or not? But the result is guaranteed to be either the fully precise fused computation rounded once, or the fully precise multiplication, rounded, then fully precise addition, rounded?

Correct.

> When I call this multiple times with the same inputs in the same program, am I guaranteed to either always get the fused behavior or always the unfused behavior, or can it differ between different invocations? What if they used different inputs?

That is an interesting question I don't know the answer to! :D

Reply:
The spec in the [LangRef for fmuladd](https://llvm.org/docs/LangRef.html#id806) suggests full-throttle nondeterminism.

Reply:
I don't know a definitive answer, but I think it's unspecified (e.g., depending on instruction latency and the surrounding code, LLVM could choose separate versus fused instructions). I haven't been able to provoke that difference in ad-hoc experiments, but that's no surprise when FMA instructions have identical latency to pure multiplication.

Reply:
Yeah the LangRef seems to make it fully non-deterministic. So we should go with that for now, and in particular document this clearly in the public function.